index.js

```diff
@@ -70,55 +87,129 @@ Worker.prototype._handleHttp = function (req, res) {
   });
 };
 
+/**
+ * Main function of the module
+ *
+ * @param req The Request of the call
+ */
 Worker.prototype.processRequest = function (req) {
-  var targetName = req.url.split('/').pop();
+  var targetName = reqToAppName(req);
   if (targetName.length === 0) return;
 
   var targetApp = this.apps[targetName];
   if (!targetApp) return;
 
-  // validate the request
+  var error = this.checkRequest(targetApp, req);
+  if (error) {
+    console.log(error);
+    return;
+  }
+
+  console.log('[%s] Received valid hook for app %s', new Date().toISOString(), targetName);
+
+  var execOptions = {
+    cwd: targetApp.cwd,
+    env: process.env,
+    shell: true
+  };
+  var phases = {
+    resolveCWD: function resolveCWD(cb) {
+      // if cwd is provided, we expect that it isnt a pm2 app
+      if (targetApp.cwd) return cb();
+
+      // try to get the cwd to execute it correctly
+      pm2.describe(targetName, function (err, apps) {
+        if (err || !apps || apps.length === 0) return cb(err || new Error('Application not found'));
+
+        // execute the actual command in the cwd of the application
+        targetApp.cwd = apps[0].pm_cwd ? apps[0].pm_cwd : apps[0].pm2_env.pm_cwd;
+        return cb();
+      });
+    },
+    pullTheApplication: function pullTheApplication(cb) {
+      vizion.update({
+        folder: targetApp.cwd
+      }, logCallback(cb, '[%s] Successfuly pulled application %s', new Date().toISOString(), targetName));
+    },
+    preHook: function preHook(cb) {
+      if (!targetApp.prehook) return cb();
+
+      spawnAsExec(targetApp.prehook, execOptions,
+          logCallback(cb, '[%s] Prehook command has been successfuly executed for app %s', new Date().toISOString(), targetName));
+    },
+    reloadApplication: function reloadApplication(cb) {
+      if (!targetApp.prehook) return cb();
```

Comment:
Ouch! I read it two times to ensure I didn't forgot nothing haha
Later today I will fix it.